### ISSUE TITLE
Fedora 34 install issue

### ISSUE THREAD

**Issue by dbutenhof** [distributed-system-analysis/pbench#2657]
Pbench agent install on Fedora 34 showed a problem due to pip explicitly refusing to *uninstall* the RPM version of `click-7.1.2` in favor of our `requirements.txt` requirement for `click >= 8.0`. 

The equivalent on RHEL 8.4 successfully uninstalls the RPM click 6 in favor of click 8. Note that we don't currently have an RPM `click` dependency for RHEL 9; would it behave like Fedora 34 if we did?

```
Installing collected packages: ifaddr, python-pidfile, hiredis, click
  Attempting uninstall: click
    Found existing installation: click 7.1.2
ERROR: Cannot uninstall click 7.1.2, RECORD file not found. Hint: The package was installed by rpm.
```

Note that the `dnf install pbench-agent` completes successfully, and the only warning visible in the `dnf` output is `pip`'s normal verbiage about installing as root (even though, in this case, we're installing into `/opt/pbench-agent/lib` as a `--prefix`). To see the problem you need to read `/opt/pbench-agent/pip3-install.log` ... or check the version of the `click` package directly.

**Comment by vishalvvr** [comment]
#### Problem 
Packages installed by rpm can be only removed/uninstalled using rpm, same is applicable for pip installations as well (else rpm db will not get updated if it can be uninstalled using pip hence not permitted).
Now real problem is latest version of `python3-click` package in f34 is `7.1.2-5.fc34` and you cannot upgrade it via rpm/dnf(no greater version is available)
```
⬢[vvijayra@toolbox Downloads]$  dnf --releasever=34 --showduplicates list python3-click
Last metadata expiration check: 0:00:01 ago on Wed Mar  2 08:58:54 2022.
Available Packages
python3-click.noarch                                                                            7.1.2-5.fc34                                                                            fedora
```
[list of build available for click in fedora linux](https://koji.fedoraproject.org/koji/packageinfo?packageID=18578) 

#### Proposed fix
1. Need to update  [pbench-agent.spec](https://github.com/distributed-system-analysis/pbench/blob/main/agent/rpm/pbench-agent.spec.j2) to use  `click >= 8.0`
i.e `Requires: python3-click >=8.0`
2. Get click version >=8.0 build for fedora 34
 - Ask fedora maintainer [Charalampos Stratakis](https://src.fedoraproject.org/user/cstratak) or [Fabian Affolter](https://src.fedoraproject.org/user/fab) of click package to update the package to latest version ( f34 is a supported release so it can be done ).
I have tried back-porting the changes of f35 into f34 and build click=8.0.4 there is no harm in do this change and/or build, only change is in %autosetup 
![Screenshot from 2022-03-02 09-17-34](https://user-images.githubusercontent.com/20382324/156299330-b4262afe-f19d-4f9a-93cc-76331c742c77.png)
- Meanwhile the changes land in official f34 as update, we can have click build in our pbench copr and fix the issue for now.
[working spec for click 8.0.4 ](https://download.copr.fedorainfracloud.org/results/vishalvvr/pbench-agent/fedora-34-x86_64/03593063-python-click/python-click.spec)

**Comment by dbutenhof** [comment]
Thanks, @vishalvvr; that's good info. As we move forward towards those Fedora packaging rules, we're going to need to replace all of our `pip` requirements with RPMs, and that might involve packaging our own ... but I feel like it'd also be better not to get ourselves into that business.

In the shorter term, maybe it'd be better to drop the RPM requirements for platforms where we know the RPM package available doesn't meet our needs, which would perhaps more simply avoid this problem. (E.g., for click, that means RHEL8.4 and Fedora, even though RHEL 8.4 doesn't currently object to the pip replacement of the RPM package.)

**Comment by vishalvvr** [comment]
I have raised an upstream PR request to make required changes [link](https://src.fedoraproject.org/rpms/python-click/pull-request/17).
This will get sorted soon, and will cause no problem once its build for f34

**Comment by webbnh** [comment]
> maybe it'd be better to drop the RPM requirements for platforms where we know the RPM package available doesn't meet our needs, which would perhaps more simply avoid this problem.

I think we're caught in a bind for which there isn't a _simple_ solution.

Our code has a specific requirement...it would be best (for external reasons) if that requirement were satisfied by an RPM package, but (for the moment), if we can satisfy it via a Pip-install, that's fine (for most of our applications).

Changing the spec file doesn't alter the situation in any positive way -- even if we don't stipulate it, our code still has that dependency.  Removing the requirement from the spec file might result in the installation completing "successfully", but without the correct version of Click in place, the result will be largely useless.

So, I think we need to come up with a way to ensure that the correct version of Click gets installed (I don't feel like we have a viable alternative...).  If `pip` and/or `dnf` won't do that for us implicitly, then I guess we need to take the appropriate steps to remove the current version in conjunction with our installation and then cause the correct version to be installed.  (This is likely to be quite interesting if there are other installed components which depend on the existing version of Click....)

As for fixing the upstream platform, I think that this is a good and positive effort to make, but I feel like it's going to be a long time before this represents an actual solution.  That is, we don't have a story on F34 until the fix comes through; once it comes through, will it automatically propagate to F35, and F36/Rawhide?  And, how do we handle RHEL 8.* and RHEL 9?  (And, what about F33 and RHEL 7?)

**Comment by webbnh** [comment]
> we're far from that goal now

Actually, for Fedora, I think we're down to about three or even a single dependency which cannot be satisfied by an RPM (now, that is, I believe, for the Agent installation, not our full development environment, but, even so, we're not "far").

> We simply have "overeager" Fedora 34 and RHEL 8 spec requirements that specify the click RPM even though it's the wrong version

Our code has a dependency on Click, so it really _should_ be in a requirements declaration in our spec file.  I thought the whole point of "distributions" was that they provided specific versions of all packages...our problem is that we need a newer version than what our favorite distributions are currently providing.  Changing our spec file won't change that.  The only ways out of that bind are to fix the distros or to find a way to install our own version of Click.

> there's another way of looking at this, as a pip defect.

I think you're on the right track, there!  However, before we chase that too far...is our installation using the most up-to-date version of Pip?  (The current version no longer suggests using `--user` when run as root if you actually specify `--user`...perhaps it also doesn't complain about RPM interactions if the target of the install is somewhere other than the system directories.)

**Comment by portante** [comment]
So far, this is really only a Fedora 34 issue, so not a blocker for the `v0.71` release.

**Comment by portante** [comment]
> Do we have any kind of reference document anywhere which lays out what what users should expect in terms of the ability to use the fruits of our labors here?

No.  We only have the COPR repos that we make available.

> Obviously, people are free to download the source code and deploy it anywhere they want, but we also have intentions to provide a better level of support for a specific set of platforms...do we have a statement as to what that set is and what that support entails?

Not that I am aware of at this point in time.

> For instance, we provide RPM packages for installation on Fedora, CentOS, and RHEL...which versions of those distributions do we expect the packages to work on? We provide Ansible playbooks which install those RPMs on some subset of those platforms. We provide containerized versions of the Agent...those work anywhere where the `podman` command runs, right?...do we think they work under `docker`? And so forth, ....

What does `podman` and `docker` have to do with this?

> Anyway, with respect to this Issue, how could I as an external user validate that (lack of) installation on Fedora 34 would not be a blocker for v0.71? E.g., do we have a definition somewhere of what comprises a release and what our release criteria are?

Why would this block this release on all the other distributions?  Why wouldn't this fix come later once we have a way to resolve it for that platform?

Most folks are using CentOS 8 and RHEL 9 at this point.  Why would I block the release for those distributions for Fedora 34?

**Comment by dbutenhof** [comment]
All good points; but we need to keep in mind the difference between upstream users who pull the GitHub repo and build/deploy on their own vs things that we, as the core team, "support". Right now that distinction can seem slightly artificial as we don't have a real product release pipeline; but we still have primary platforms we know our main users need and which we test and support. I think maybe Peter's point is that Fedora 34 is not on that list; and even though we expect that it'll work, we don't necessarily claim formal 0.71 "support"?

**Comment by webbnh** [comment]
> we still have primary platforms we know our main users need and which we test and support. I think maybe Peter's point is that Fedora 34 is not on that list

Right...so, I was wondering what _is_ on that list.  (I.e., what _would_ we consider the release blocked for?)
